packages/async-flow/src/template.js

```diff
@@ -0,0 +1,130 @@
+/**
+ * @import {HostCall} from './types.js';
+ */
+
+const { assign, defineProperty } = Object;
+
+/** Doesn't need to be exhaustive, just a little prettier than JSON-quoting. */
+const BEST_GUESS_ID_REGEX = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/;
+
+/**
+ * Return an object that mimics a template strings array.
+ *
+ * @param {string[]} strings
+ * @param {string[]} [raw] optional raw strings, defaults to `strings`
+ * @returns {TemplateStringsArray}
+ */
+export const makeTemplateStringsArray = (strings, raw = strings) => {
+  // Get the shape of the result object to match TemplateStringsArray.
+  const result = assign([...strings], { raw });
+
+  // Make the raw property non-enumerable.
+  defineProperty(result, 'raw', {
+    value: raw,
+    enumerable: false,
+  });
+
+  return harden(result);
+};
+harden(makeTemplateStringsArray);
+
+/**
+ * When used as a template tag, this function returns its arguments verbatim.
+ *
+ * @template {any[]} A
+ * @param  {A} allArgs
+ * @returns {A}
+ */
+export const idTemplateTag = (...allArgs) => allArgs;
+harden(idTemplateTag);
+
+/**
+ * Convert a replay membrane HostCall structure to template arguments.
+ *
+ * @param {HostCall} hostCall
+ * @returns {[TemplateStringsArray, ...any[]]}
+ */
+export const hostCallToTemplateArgs = ({ target, method, eventual }) => {
+  /** @type {string[]} */
+  const tmpl = [];
+
+  /** @type {any[]} */
+  const args = [];
+
+  const tpush = str => {
+    tmpl.push(str);
+  };
+  const tappend = str => {
+    tmpl[tmpl.length - 1] += str;
+  };
+
+  tpush(eventual ? 'E' : '');
+  tappend('(');
+  args.push(target);
+  tpush(')');
+  if (typeof method === 'string') {
+    if (BEST_GUESS_ID_REGEX.test(method)) {
+      tappend(`.${method}`);
+    } else {
+      tappend(`[${JSON.stringify(method)}]`);
+    }
+  } else if (method !== undefined) {
+    tappend(`[${String(method)}]`);
+  }
+  tappend('(...)');
+
+  return /** @type {const} */ ([makeTemplateStringsArray(tmpl), ...args]);
+};
+harden(hostCallToTemplateArgs);
+
+/**
+ * Template tag to flatten any nested template arguments by joining them to the
+ * returned template strings and rest arguments.
+ *
+ * @param {TemplateStringsArray} tmpl
+ * @param  {any[]} args
+ * @returns {[TemplateStringsArray, ...any[]]}
+ */
+export const inlineTemplateArgs = (tmpl, ...args) => {
```

Comment:
I am wondering if the ergonomics wouldn't be simpler with a wrapper for the original template tag function:

```js
inlineArgs(X)`foo bar ${inlinedArg}`;
```